tests/integration/mongos/ldap/test_ldap.py

```diff
@@ -0,0 +1,215 @@
+
+from pathlib import Path
+
+import pytest
+from juju.model import Model
+from pytest_operator.plugin import OpsTest
+
+from ...helpers import (
+    DATA_INTEGRATOR_APP_NAME,
+    DEPLOYMENT_TIMEOUT,
+    check_or_scale_app,
+    deploy_charm,
+    execute_on_mongod,
+    get_app_name,
+    wait_for_mongodb_units_blocked,
+)
+from ...ldap_helpers import (
+    LDAP_CERT_OFFER,
+    LDAP_OFFER,
+    apply_ldif,
+    consume_offers,
+    create_groups,
+    deploy_glauth,
+    generate_mongodb_ldap_client,
+    teardown_offers,
+)
+from ...sharding_helpers import (
+    CLUSTER_COMPONENTS,
+    CLUSTER_REL_NAME,
+    CONFIG_SERVER_APP_NAME,
+    SHARD_ONE_APP_NAME,
+    SHARD_TWO_APP_NAME,
+    deploy_cluster_components,
+    integrate_cluster,
+)
+
+TIMEOUT = 15 * 60
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy_mongodb_cluster(
+    ops_test: OpsTest,
+    mongodb_charm: Path,
+    substrate: str,
+    mongod_resource,
+    kubernetes_model: Model,
+) -> None:
+    """Build and deploy one unit of MongoDB."""
+    # it is possible for users to provide their own cluster for testing. Hence check if there
+    # is a pre-existing cluster.
+    await deploy_cluster_components(
+        ops_test,
+        substrate=substrate,
+        mongodb_charm=mongodb_charm,
+        mongod_resource=mongod_resource,
+        extra_config_config_server={
+            "ldap-query-template": "dc=glauth,dc=com??sub?(&(objectClass=posixGroup)(uniqueMember={PROVIDED_USER}))"
+        },
+    )
+    await ops_test.model.wait_for_idle(
+        apps=CLUSTER_COMPONENTS,
+        idle_period=20,
+        timeout=DEPLOYMENT_TIMEOUT,
+        raise_on_blocked=False,
+    )
+    await integrate_cluster(ops_test)
+    await ops_test.model.wait_for_idle(
+        apps=CLUSTER_COMPONENTS,
+        status="active",
+        idle_period=20,
+        timeout=DEPLOYMENT_TIMEOUT,
+    )
+
+    # deploy the glauth-k8s charm
+    await deploy_glauth(ops_test, kubernetes_model)
+    await consume_offers(ops_test, kubernetes_model)
+
+    db_app_name = CONFIG_SERVER_APP_NAME
+
+    await create_groups(
+        ops_test, substrate, db_app_name, "ou=superheroes,ou=users,dc=glauth,dc=com"
+    )
+
+    await apply_ldif(ops_test, kubernetes_model, "add.ldif")
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy_mongos(
+    ops_test: OpsTest, mongos_charm: Path, substrate: str, mongod_resource, base_app_name
+) -> None:
+    if app_name := await get_app_name(ops_test, ch_name="mongos"):
+        await check_or_scale_app(ops_test, app_name, 3)
+    else:
+        await deploy_charm(
+            ops_test=ops_test,
+            charm=mongos_charm,
+            substrate=substrate,
+            mongod_resource=mongod_resource,
+            app_name=base_app_name,
+            num_units=3,
+            subordinate=(substrate == "lxd"),
+        )
+        app_name = base_app_name
+
+    await ops_test.model.deploy(
+        DATA_INTEGRATOR_APP_NAME,
```

Comment:
This is excusively needed for sharding on VM, no?

Reply:
Yes, but it is easier for those tests to follow the exact same workflow.
I'm still working on how to make those tests extremely clean, and how we will have different workflows for different substrates.
In that case, it doesn't change what we test so I'd rather keep it (and even more if we create roles with data integrator in the future, this will already be there).

Reply:
TODO: Add comment for posterity.